src/osd/scrubber/scrub_machine.cc

```diff
@@ -109,83 +109,140 @@ sc::result NotActive::react(const AfterRepairScrub&)
-  scrbr->reserve_replicas();
-
-  auto timeout = scrbr->get_cct()->_conf.get_val<
-    std::chrono::milliseconds>("osd_scrub_reservation_timeout");
-  if (timeout.count() > 0) {
-    // Start a timer to handle case where the replicas take a long time to
-    // ack the reservation.  See ReservationTimeout handler below.
-    m_timeout_token = machine.schedule_timer_event_after<ReservationTimeout>(
-      timeout);
+  // initiate the reservation process
+  context<Session>().m_reservations.emplace(*scrbr);
+
+  if (context<Session>().m_reservations->get_last_sent()) {
+    // the 1'st reservation request was sent
+
+    auto timeout = scrbr->get_pg_cct()->_conf.get_val<milliseconds>(
+	"osd_scrub_reservation_timeout");
+    if (timeout.count() > 0) {
+      // Start a timer to handle case where the replicas take a long time to
+      // ack the reservation.  See ReservationTimeout handler below.
+      m_timeout_token =
+	  machine.schedule_timer_event_after<ReservationTimeout>(timeout);
+    }
+  } else {
+    // no replicas to reserve
+    dout(10) << "no replicas to reserve" << dendl;
+    // can't transit directly from here
+    post_event(RemotesReserved{});
   }
 }
 
 ReservingReplicas::~ReservingReplicas()
 {
   DECLARE_LOCALS;  // 'scrbr' & 'pg_id' aliases
-  if (m_holding_isreserving_flag) {
-    scrbr->clear_reserving_now();
-  }
+  // it's OK to try and clear the flag even if we don't hold it
+  // (the flag remembers the actual holder)
+  scrbr->clear_reserving_now();
 }
 
-sc::result ReservingReplicas::react(const ReservationTimeout&)
+sc::result ReservingReplicas::react(const ReplicaGrant& ev)
 {
   DECLARE_LOCALS;  // 'scrbr' & 'pg_id' aliases
-  dout(10) << "ReservingReplicas::react(const ReservationTimeout&)" << dendl;
+  dout(10) << "ReservingReplicas::react(const ReplicaGrant&)" << dendl;
 
-  const auto msg = fmt::format(
-      "PgScrubber: {} timeout on reserving replicas (since {})",
-      scrbr->get_spgid(), entered_at);
-  dout(5) << msg << dendl;
-  scrbr->get_clog()->warn() << "osd." << scrbr->get_whoami() << " " << msg;
-  scrbr->on_replica_reservation_timeout();
+  if (context<Session>().m_reservations->handle_reserve_grant(
+	  ev.m_op, ev.m_from)) {
+    // we are done with the reservation process
+    return transit<ActiveScrubbing>();
+  }
   return discard_event();
 }
 
-sc::result ReservingReplicas::react(const ReservationFailure&)
+sc::result ReservingReplicas::react(const ReplicaReject& ev)
 {
   DECLARE_LOCALS;  // 'scrbr' & 'pg_id' aliases
-  dout(10) << "ReservingReplicas::react(const ReservationFailure&)" << dendl;
+  dout(10) << "ReservingReplicas::react(const ReplicaReject&)" << dendl;
 
-  // the Scrubber must release all resources and abort the scrubbing
-  scrbr->clear_pgscrub_state();
+  // manipulate the 'next to reserve' iterator to exclude
+  // the rejecting replica from the set of replicas requiring release
+  context<Session>().m_reservations->verify_rejections_source(
```

Comment:
Seems like a very misleading name.  How about notify_reservation_rejected()?

Reply:
The name matches the functionality as after the rest of the commits (where the notification part was
moved to the FSM). But I will remove the extra 's' there.